### PR TITLE
refactor(positioning): refactor positioning service

### LINKED ISSUE #2796: fix(popover)(tooltip): "auto" position is not working correctly
**Feature**

`"auto"` positioning in the `[popover]` directive. This could be related to issue #2661, but I'm not sure.

**Bugs**

1. The popover goes off the top of the screen by about 100 pixels before it moves to the `"bottom"` position.
2. When the popover would go over the right side of the screen, it always resorts to the `"top"` position.

**Context**

version: `1.9.3`

The application is a dashboard with draggable tiles, each having a popover in the top right corner.

This is the current configuration for the popover:

- `[popover]="popoverTemplate"`
- `placement="auto"`
- `container="body"` (several parent elements have `overflow: hidden`)
- `[outsideClick]="true"`

I have created a simplified demo here: https://github.com/nfiles/ngx-bootstrap-popover-issue-demo

**Misc**

I wasn't sure where to go next in my debugging, so I started to look at the source. In `src/positioning/ng-positioning.ts` at line 197 I noticed that if the `"top"` and `"bottom"` else-if blocks were switched, everything worked as expected. I assume this is because `"top"` is preferred over `"bottom"`, but I do not know why the `"top"` position is chosen when the popover is going off screen.

### LINKED ISSUE #2993: fix(datepicker): lag on datepicker in a modal re-alignment after scrolling
### Bug description or feature request:

If you have a datepicker in a modal, and then scroll the modal the datepicker is left behind. After a long pause (maybe 10 seconds), the datepicker will jump back to the input. If you click the datepicker that also causes it to jump to the input.

### Versions of ngx-bootstrap, Angular, and Bootstrap:

ngx-bootstrap: 1.9.3

Angular: 4.3.2

Bootstrap: bootstrap-sass 3.3.7

### Build system: Angular CLI, System.js, webpack, starter seed:

Webpack
[yo angular2-typescript generator](https://github.com/shibbir/generator-angular2-typescript) version 0.8.1.

### LINKED ISSUE #3303: fix(popover): all popovers, that are placed near left side of demo are cutted by viewport borders, if left side menu is not shown
### Bug description:
- Open demo on mobile device or on desktop with window width 768px or less (to hide left side navigation menu)
- Go to popover demo page
- Open any popover, that will be placed near left border of viewport
### Expected result:
All demo elements should be fully displayed

popoveroutsideviewport
### Versions of Angular, and Bootstrap:

Angular: all supported

Bootstrap: all supported

### LINKED ISSUE #4470: bsDatepicker should align right below the text-box.
I want the date picker to align bottom right of text box. But when I use placement="bottom right" the datepicker starts from end of the text box. Where as the datepicker should end below the textbox and should not go beyond the textbox.
currently i am getting datepicker as below.

Need something as below.

### PR DESCRIPTION
**Pull request by Domainv**
Closes #3303
Closes #2993
Closes #4470
Closes #2796
# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated tests.
 - [ ] added/updated API documentation.
 - [ ] added/updated demos.

### CODE REVIEW

File: src/positioning/ng-positioning.ts [hunk 1]
```diff
@@ -256,13 +53,16 @@ export function positionElements(
...
     placement,
     appendToBody
   );
 
-  targetElement.style.top = `${pos.top}px`;
-  targetElement.style.left = `${pos.left}px`;
+  targetElement.style['will-change'] = 'transform';
+  targetElement.style.top = '0px';
+  targetElement.style.left = '0px';
+  targetElement.style.transform = `translate3d(${Math.floor(pos.left)}px, ${Math.floor(pos.top)}px, 0px)`;
```
Comment (Domainv) [inline src/positioning/ng-positioning.ts]: remove Math.floor

**Comment by kulenkov** [comment]
@Domainv ,
checked next components: Popover, Tooltip, Typeahead  - looks good in desktop on the next browsers and resolutions:
browsers: chrome v72, safari v12
chrome with the next resolutions 1920x1080, 1024x786, 800x600
safari 2560x1600 
there is an issue on mobile for Tooltip and Popover: 
tooltip_placment


Also: for Typeahead / DropUp - the list should be displayed upper the input:
dropup

**Comment by kulenkov** [comment]
@Domainv ,
it seems that the Datepicker become broken. It doesn't appear now by clicking in the inputs or on the buttons, please see the attached video:
datepicker

**Comment by kulenkov** [comment]
@Domainv ,
Component: Dropdown / Demo: Nested Dropdowns
Doesn't show by clicking - it's actual for all browsers. please see the video below:
dropdowns_nested

Also I've found the issue for Safari:
Component: Dropdown/ Demo: Append to body - the dropdown doesn't show by clicking
please see the video below:
dropdown_safari

**Comment by kulenkov** [comment]
checked after fixing - looks good to me

**Comment by kulenkov** [comment]
@Domainv ,
tested. on safari, chrome, mobile: safari. looks good to me!

**Comment by kulenkov** [comment]
re-checked after last changes: looks good to me

**Pull request merged**